File: lecture09/exercise09-D/template.py
#!/usr/bin/env python3

# Exercise 09-D: Invert Binary Tree

from collections import deque
from dataclasses import dataclass
from typing      import Optional
import sys

# Classes

@dataclass
class Node:
    value:  int
    left:   Optional['Node']
    right:  Optional['Node']

# Functions

def read_tree(values: list[int], index: int=0) -> Optional[Node]:
    # TODO: Use divide and conquer to parse tree
    return None

def walk_tree(root: Optional[Node]) -> list[int]:
    # TODO: Use BFS Traversal with Queue to create list of values
    return []

def dump_tree(root: Optional[Node]) -> None:
    print(','.join(map(str, walk_tree(root))))

def invert_tree(root: Optional[Node]) -> Optional[Node]:
    # TODO: Use divide and conquer to invert binary tree

    # Base Case: Invalid Node

    # Divide and Conquer: Swap left and right and recurse
    return None

# Main Execution

def main() -> None:
    for line in sys.stdin:
        values = list(map(int, line.split()))
        root   = read_tree(values)

        invert_tree(root)
        dump_tree(root)

if __name__ == '__main__':
    main()

File: lecture09/exercise09-D/template.cpp
// Exercise 09-D: Invert Binary Tree

#include <iostream>
#include <memory>
#include <queue>
#include <string>
#include <sstream>
#include <vector>

using namespace std;

// Node structures

template <typename T>
struct Node {
    T value;
    struct Node *left;
    struct Node *right;
};

typedef Node<int> IntNode;

// Functions

bool read_values(vector<int> &values) {
    values.clear();

    string line;
    if (getline(cin, line)) {
    	stringstream ss(line);
    	int i;

    	while (ss >> i) {
    	    values.push_back(i);
	}
    }

    return !values.empty();
}

IntNode*  read_tree(vector<int> &values, size_t index=0) {
    return nullptr;
}

void	dump_tree(IntNode *root) {
}

IntNode*    invert_tree(IntNode *root) {
    // Base Case: Invalid Node

    // Divide and Conquer: Swap left and right and recurse
    return nullptr;
}

// Main execution

int main(int argc, char *argv[]) {
    vector<int> values;

    while (read_values(values)) {
    	auto root = read_tree(values);
    	invert_tree(root);
    	dump_tree(root);
    }

    return 0;
}
